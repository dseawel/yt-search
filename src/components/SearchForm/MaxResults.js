import React from 'react';
import PropTypes from 'prop-types';
import {FormControl, InputLabel, MenuItem, Select} from '@material-ui/core';
import {withStyles} from '@material-ui/core/styles';

const propTypes = {
  classes: PropTypes.object,
  handlesetMaxResults: PropTypes.func,
  maxResults: PropTypes.string,
};

const styles = {
  maxResultsSelect: {
    width: 100,
  },
};

const MaxResults = ({classes, handlesetMaxResults, maxResults}) => (
  <FormControl>
    <InputLabel htmlFor="max">{'Max Results: '}</InputLabel>
    <Select
      id="max"
      onChange={value => handlesetMaxResults(value.target.value)}
      value={maxResults}
      className={classes.maxResultsSelect}
    >
      <MenuItem value={'1'}>{'1'}</MenuItem>
      <MenuItem value={'5'}>{'5'}</MenuItem>
      <MenuItem value={'10'}>{'10'}</MenuItem>
    </Select>
  </FormControl>
);

MaxResults.propTypes = propTypes;

export default withStyles(styles)(MaxResults);
