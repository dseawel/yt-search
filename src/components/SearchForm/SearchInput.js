import React from 'react';
import PropTypes from 'prop-types';
import {
  FormControl,
  Input,
  InputLabel,
  InputAdornment,
  Typography,
} from '@material-ui/core';
import SearchIcon from '@material-ui/icons/Search';

const propTypes = {
  error: PropTypes.string,
  handlesetSearchParam: PropTypes.func,
};

const SearchInput = props => (
  <FormControl>
    <InputLabel htmlFor="adornment-search">{'Search YouTube'}</InputLabel>
    <Input
      id="adornment-search"
      startAdornment={
        <InputAdornment position="start">
          <SearchIcon />
        </InputAdornment>
      }
      placeholder="Keyword"
      onChange={searchParam =>
        props.handlesetSearchParam(searchParam.target.value)
      }
    />
    {props.error && <Typography component="em">{props.error}</Typography>}
  </FormControl>
);

SearchInput.propTypes = propTypes;

export default SearchInput;
