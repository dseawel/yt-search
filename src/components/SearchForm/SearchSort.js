import React from 'react';
import PropTypes from 'prop-types';
import {FormControl, InputLabel, MenuItem, Select} from '@material-ui/core';

const propTypes = {
  handlesetSearchSort: PropTypes.func,
  searchSort: PropTypes.string,
};

const SearchSort = props => (
  <FormControl>
    <InputLabel htmlFor="sort">{'Sort by: '}</InputLabel>
    <Select
      id="sort"
      onChange={searchSort =>
        props.handlesetSearchSort(searchSort.target.value)
      }
      value={props.searchSort}
    >
      <MenuItem value={'date'}>{'Date'}</MenuItem>
      <MenuItem value={'rating'}>{'Rating'}</MenuItem>
      <MenuItem value={'relevance'}>{'Relevance'}</MenuItem>
    </Select>
  </FormControl>
);

SearchSort.propTypes = propTypes;

export default SearchSort;
