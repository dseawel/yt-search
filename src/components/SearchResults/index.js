import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import SearchResultItem from './SearchResultItem';

const propTypes = {
  searchResults: PropTypes.arrayOf(PropTypes.object),
};

class SearchResults extends PureComponent {
  static propTypes = {
    searchResults: PropTypes.arrayOf(PropTypes.object),
  };

  render() {
    const {searchResults} = this.props;

    if (!searchResults) {
      return <p>{'No results to display.'}</p>;
    }

    return searchResults.map(searchResultItem => (
      <SearchResultItem
        key={searchResultItem.etag}
        searchResultItem={searchResultItem}
      />
    ));
  }
}

SearchResults.propTypes = propTypes;

const mapStateToProps = ({ytSearch: {searchResults}}) => {
  return {searchResults};
};

export default connect(
  mapStateToProps,
  {}
)(SearchResults);
