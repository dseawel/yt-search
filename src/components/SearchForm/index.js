import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {
  fetchYouTubeSearchResults,
  setMaxResults,
  setSearchParam,
  setSearchSort,
} from 'actions/yt-search';
import SearchInput from './SearchInput';
import SearchSort from './SearchSort';
import MaxResults from './MaxResults';
import {Button, CircularProgress} from '@material-ui/core';

export class SearchForm extends PureComponent {
  static propTypes = {
    error: PropTypes.object,
    fetchYouTubeSearchResults: PropTypes.func,
    maxResults: PropTypes.string,
    setMaxResults: PropTypes.func,
    searchParam: PropTypes.string,
    setSearchParam: PropTypes.func,
    searchSort: PropTypes.string,
    setSearchSort: PropTypes.func,
  };

  constructor(props) {
    super(props);

    this.state = {
      displayErrors: false,
      isValid: false,
      loading: false,
      validationErrors: {},
    };
  }

  _handleSearch = async () => {
    await this._validateForm();
    if (this.state.isValid) {
      this.setState({loading: true});
      this.props
        .fetchYouTubeSearchResults(this.props)
        .then(() => {
          this.setState({loading: false});
        })
        .catch(error => {
          alert('Error from YouTube');
          console.log('error', error);
        });
    }
  };

  _validateForm = () => {
    const {searchParam} = this.props;
    const validationErrors = {};
    if (!searchParam || searchParam.length < 2) {
      validationErrors.searchParam = 'Search keyword is missing';
    }
    this.setState({
      validationErrors,
      isValid: Object.keys(validationErrors).length === 0,
    });
  };

  _handlesetSearchParam = searchParam => {
    this.props.setSearchParam({searchParam});
  };

  _handlesetSearchSort = searchSort => {
    this.props.setSearchSort({searchSort});
  };

  _handlesetMaxResults = maxResults => {
    this.props.setMaxResults({maxResults});
  };

  render() {
    return (
      <div>
        <div>
          <SearchInput
            handlesetSearchParam={this._handlesetSearchParam}
            error={this.state.validationErrors.searchParam}
          />
          &nbsp;
          <MaxResults
            handlesetMaxResults={this._handlesetMaxResults}
            maxResults={this.props.maxResults}
          />
          &nbsp;
          <SearchSort
            handlesetSearchSort={this._handlesetSearchSort}
            searchSort={this.props.searchSort}
          />
        </div>
        <br />
        <div>
          <Button
            size="medium"
            color="primary"
            variant="contained"
            onClick={this._handleSearch}
          >
            {this.state.loading ? (
              <CircularProgress color="inherit" size={20} />
            ) : (
              'Search'
            )}
          </Button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ytSearch}) => {
  return {...ytSearch};
};

export default connect(
  mapStateToProps,
  {
    fetchYouTubeSearchResults,
    setMaxResults,
    setSearchParam,
    setSearchSort,
  }
)(SearchForm);
