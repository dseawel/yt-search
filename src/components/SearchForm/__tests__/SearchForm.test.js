import React from 'react';
import renderer from 'react-test-renderer';
import {SearchForm} from 'components/SearchForm';

test('SearchForm renders without crashing', () => {
  let props = {
    error: {},
    fetchYouTubeSearchResults: jest.fn(),
    maxResults: '1',
    setMaxResults: jest.fn(),
    searchParam: null,
    setSearchParam: jest.fn(),
    searchSort: 'rating',
    setSearchSort: jest.fn(),
  };
  const emptyComponent = <SearchForm {...props} />;
  let rendered = renderer.create(emptyComponent).toJSON();
  expect(rendered).toMatchSnapshot();

  //test with updated search criteria
  props = {
    error: {},
    fetchYouTubeSearchResults: jest.fn(),
    maxResults: '10',
    setMaxResults: jest.fn(),
    searchParam: 'Jack Henry',
    setSearchParam: jest.fn(),
    searchSort: 'date',
    setSearchSort: jest.fn(),
  };
  const updatedComponent = <SearchForm {...props} />;
  rendered = renderer.create(updatedComponent).toJSON();
  expect(rendered).toMatchSnapshot();
});
