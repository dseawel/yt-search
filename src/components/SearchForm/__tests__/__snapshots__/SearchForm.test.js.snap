// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders without crashing 1`] = `
<div>
  <div>
    <div
      className="MuiFormControl-root-1"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
        htmlFor="adornment-search"
      >
        Search YouTube
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-adornedStart-40"
        onClick={[Function]}
      >
        <div
          className="MuiInputAdornment-root-53 MuiInputAdornment-positionStart-55"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-58"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputAdornedStart-51"
          disabled={false}
          id="adornment-search"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Keyword"
          required={false}
          type="text"
        />
      </div>
    </div>
     
    <div
      className="MuiFormControl-root-1"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
        htmlFor="max"
      >
        Max Results: 
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24 MaxResults-maxResultsSelect-67"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root-68"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            className="MuiSelect-select-69 MuiSelect-selectMenu-72 MuiInputBase-input-46 MuiInput-input-31"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            1
          </div>
          <input
            id="max"
            type="hidden"
            value="1"
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-58 MuiSelect-icon-74"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
     
    <div
      className="MuiFormControl-root-1"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
        htmlFor="sort"
      >
        Sort by: 
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root-68"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            className="MuiSelect-select-69 MuiSelect-selectMenu-72 MuiInputBase-input-46 MuiInput-input-31"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            Rating
          </div>
          <input
            id="sort"
            type="hidden"
            value="rating"
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-58 MuiSelect-icon-74"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <br />
  <div>
    <button
      className="MuiButtonBase-root-105 MuiButton-root-79 MuiButton-contained-90 MuiButton-containedPrimary-91 MuiButton-raised-93 MuiButton-raisedPrimary-94"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-80"
      >
        Search
      </span>
      <span
        className="MuiTouchRipple-root-108"
      />
    </button>
  </div>
</div>
`;

exports[`SearchForm renders without crashing 2`] = `
<div>
  <div>
    <div
      className="MuiFormControl-root-1"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
        htmlFor="adornment-search"
      >
        Search YouTube
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-adornedStart-40"
        onClick={[Function]}
      >
        <div
          className="MuiInputAdornment-root-53 MuiInputAdornment-positionStart-55"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-58"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputAdornedStart-51"
          disabled={false}
          id="adornment-search"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Keyword"
          required={false}
          type="text"
        />
      </div>
    </div>
     
    <div
      className="MuiFormControl-root-1"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
        htmlFor="max"
      >
        Max Results: 
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24 MaxResults-maxResultsSelect-67"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root-68"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            className="MuiSelect-select-69 MuiSelect-selectMenu-72 MuiInputBase-input-46 MuiInput-input-31"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            10
          </div>
          <input
            id="max"
            type="hidden"
            value="10"
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-58 MuiSelect-icon-74"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
     
    <div
      className="MuiFormControl-root-1"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
        htmlFor="sort"
      >
        Sort by: 
      </label>
      <div
        className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-formControl-37 MuiInput-formControl-24"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root-68"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            className="MuiSelect-select-69 MuiSelect-selectMenu-72 MuiInputBase-input-46 MuiInput-input-31"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            Date
          </div>
          <input
            id="sort"
            type="hidden"
            value="date"
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-58 MuiSelect-icon-74"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <br />
  <div>
    <button
      className="MuiButtonBase-root-105 MuiButton-root-79 MuiButton-contained-90 MuiButton-containedPrimary-91 MuiButton-raised-93 MuiButton-raisedPrimary-94"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-80"
      >
        Search
      </span>
      <span
        className="MuiTouchRipple-root-108"
      />
    </button>
  </div>
</div>
`;
